File: python/sqs_consumer/consumer.py
import os
import time
import json
import boto3
from dotenv import load_dotenv

load_dotenv()


boto3.setup_default_session(profile_name=os.environ.get("AWS_PROFILE"))
sqs = boto3.resource(
    "sqs",
    region_name=os.environ.get("AWS_REGION"),
    endpoint_url=os.environ.get("AWS_ENDPOINT"),
)
queue_name = os.environ.get("AWS_QUEUE")
try:
    queue = sqs.get_queue_by_name(QueueName=queue_name)
    print("+++ boot python consumer.\n")
except Exception as e:
    params = {
        "message": str(e),
        "queue": queue_name,
    }
    print("localstackが起動していないか、キューが存在していないみたいです. {}".format(json.dumps(params)))
    exit(1)


def consume(body: str):
    print("python has received sqs message. : {}".format(body))


if __name__ == "__main__":
    while True:
        messages = queue.receive_messages()
        for message in messages:
            consume(message.body)
            message.delete()
        time.sleep(1)

File: js/consumer.ts
import { SQSClient, ReceiveMessageCommand, DeleteMessageCommand, GetQueueUrlCommand } from "@aws-sdk/client-sqs"
import * as dotenv from 'dotenv'

dotenv.config()

const options = {
  profile: process.env.AWS_PROFILE,
  region: process.env.AWS_REGION,
  endpoint: process.env.AWS_ENDPOINT_URL,
}
const queue = process.env.AWS_QUEUE || "none"

const client = new SQSClient(options)

const getQueueUrl = (queue: string) =>
  client.send(
    new GetQueueUrlCommand({
      QueueName: queue
    })
  )

const receiveMessage = (queueUrl: string) =>
  client.send(
    new ReceiveMessageCommand({
      MaxNumberOfMessages: 1,
      MessageAttributeNames: ["All"],
      QueueUrl: queueUrl
    })
  )

const sleep = (msec: number) => new Promise(resolve => {
  setTimeout(resolve, msec)
})

export const consume = async (queue: string) => {
  let queue_url: any = "none"
  try {
    queue_url = await (await getQueueUrl(queue)).QueueUrl
    console.log("+++ boot js/ts consumer.\n")
  } catch (e: any) {
    const param = {
      "message": e.message,
      "queue": queue,
    }
    console.log("localstackが起動していないか、キューが存在していないみたいです. " + JSON.stringify(param))
    process.exit(1)
  }
  while (true) {
    const { Messages } = await receiveMessage(queue_url);
    if (Messages) {
      console.log("Js/Ts has received sqs message. : " + Messages[0].Body)
      await client.send(
        new DeleteMessageCommand({
          QueueUrl: queue_url,
          ReceiptHandle: Messages[0].ReceiptHandle
        })
      )
    }
    await sleep(1000);
  }
}
consume(queue)
